### DISCUSSION TITLE
Should code be marked as unreachable if an attribute marked as `Never` is accessed?

### DISCUSSION THREAD

**Discussion by Viicos** [microsoft/pyright#7147] [Q&A] [answered]
pyright understands the following (i.e. it marks `b = 1` as being unreachable):

Code sample in [pyright playground](https://pyright-play.net/?pyrightVersion=1.1.328&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoByYASgKYwCuIKANIaQG6kgBQbAxgDYCGAzr1ACCALhZRxUACalgUBOAQAKXqU7AAlFAC0APkIlyVFKImmoAOkttuUALxDF68-LBL1bAEZ2oARiA)

```python
from typing import NoReturn, Never

class A:
    def prop(self) -> NoReturn:
        ...

a = A().prop()

b = 1
```

However, it does not with a property (or any attribute marked as `Never`/`NoReturn`):

Code sample in [pyright playground](https://pyright-play.net/?pyrightVersion=1.1.328&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoByYASgKYwCuIKANIaQG6kgBQbAxgDYCGAzr1ACCALjZRxUAAIJwCZvBYSoAE1LAoMsAgAUvUp2ABKKAFoAfIRLkqKUUqUA6J225QAvEO2GHmhGwBG7lAAjEA)

```python
from typing import NoReturn, Never

class A:

    @property
    def prop(self) -> NoReturn:
        ...

a = A().prop

b = 1
```

The spec only [specifies](https://typing.readthedocs.io/en/latest/spec/special-types.html#noreturn) function calls, and I'm wondering if attributes should be taken into account. Right now, pyright does not emit any diagnostic when making use of a type annotated as `NoReturn`/`Never`. Are the current semantics correct? Or could the spec be expanded to handle such scenarios? Maybe only for properties?

Thanks in advance

**Comment by erictraut** [comment, marked as answer]
This is by design. The spec says nothing about reachability detection and reporting. That's something that type checkers can choose to implement (or not) any way they see fit. Pyright's reachability detection takes into account only explicit calls that return `NoReturn`. It doesn't take into account implicit magic method invocations. Adding such support would be exorbitantly expensive at analysis time, so it wouldn't be a good tradeoff.

**Reply by Viicos** [reply]
Thanks for the quick reply, do you think explicitly stating this information in the spec would make sense? If so I can send a patch (for reference, it was explicitly stated in [PEP 484](https://peps.python.org/pep-0484/#the-noreturn-type), so maybe it was removed for a reason)

**Reply by erictraut** [reply]
This is an area where I don't think the spec should mandate type checker behaviors, so I would oppose modifying the spec in this manner.

**Reply by Viicos** [reply]
Sorry by _this information_ I meant "valid locations for `NoReturn`", PEP 484 states:

> The `NoReturn` type is only valid as a return annotation of functions, and considered an error if it appears in other positions

But I guess it was removed from the spec so that type checkers can actually implement support for other locations, as you said

**Reply by erictraut** [reply]
Ah, thanks for the clarification. Yes, the spec was recently modified to officially allow `NoReturn` (and its alias `Never`) anywhere. This decision was approved by the Typing Council [here](https://github.com/python/typing-council/issues/5) after a discussion in [this forum](https://github.com/python/typing/issues/1558).
